src/sparseml/transformers/compression/compressors/sparse_bitmask.py

```diff
@@ -0,0 +1,195 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import Dict, Tuple
+
+import numpy
+import torch
+from torch import Tensor
+from tqdm import tqdm
+
+from sparseml.transformers.compression.compressors import ModelCompressor
+
+
+__all__ = [
+    "BitmaskCompressor",
+    "BitmaskTensor",
+    "bitmask_compress",
+    "bitmask_decompress",
+    "pack_bitmasks",
+    "unpack_bitmasks",
+]
+
+_LOGGER: logging.Logger = logging.getLogger(__name__)
+
+
+@ModelCompressor.register(name="sparse_bitmask")
+class BitmaskCompressor(ModelCompressor):
+    """
+    Compression for sparse models using bitmasks. Non-zero weights are stored in a 1d
+    values tensor, with their locations stored in a 2d bitmask
+    """
+
+    def compress(self, model_state: Dict[str, Tensor]) -> Dict[str, Tensor]:
+        """
+        Compresses a dense state dict using bitmask compression
+
+        :param model_state: state dict of uncompressed model
+        :return: compressed state dict
+        """
+        compressed_dict = {}
+        _LOGGER.info(
+            f"Compressing model with {len(model_state)} parameterized layers..."
+        )
+        for name, value in tqdm(model_state.items()):
+            bitmask_tensor = BitmaskTensor(value)
+            compressed_dict |= bitmask_tensor.dict(name_prefix=name)
+
+        return compressed_dict
+
+    def decompress(self, model_state: Dict[str, Tensor]) -> Dict[str, Tensor]:
+        """
+        Uncompresses a bitmask compressed state dict back to dense
+
+        :param model_state: state dict of uncompressed model
+        :return: compressed state dict
+        """
+        raise NotImplementedError()
+
+
+class BitmaskTensor:
+    """
+    Owns compressions and decompression for a single bitmask compressed tensor.
+    Adapted from: https://github.com/mgoin/torch_bitmask/tree/main
+
+    :param tensor: Dense tensor to compress
+    """
+
+    def __init__(self, tensor: Tensor):
+        self.shape = tensor.shape
+        self.values, self.bitmasks, self.row_offsets = bitmask_compress(tensor.cpu())
```

Comment:
should we move these compressed tensors to the original device of tensor? currently they are always going to be on the cpu after this line, right?

Reply:
I think for compression this is desired. I'm assuming we want to keep the dense model in GPU memory, so rather than doubling the GPU memory we just offload to CPU. While its easy to load safetensor files tensor by tensor, I don't think there is a way to do this on save (unless I'm missing something)

Reply:
That being said though, I think it makes sense to at least save the dense tensor device so we can use it on decompression. Added that in to my latest commit

Reply:
Okay that makes sense, thanks